r/tests/testthat/test-pseudobulk.R

```diff
@@ -0,0 +1,147 @@
+      expect_equal(m_sum, m_bpcells_sum)
+      # check for clipping if matrix isn't transposed
+      if (m_bpcells@transpose == FALSE) {
+        for (quantile in c(0.1, 0.25, 0.5, 0.75, 0.99)) {
+          min_vals  <- m %>% apply(2, quantile, probs = quantile, type = 7) %>% unlist()
+          m_clipped <- t(pmin(t(m), min_vals))
+          m_sum_clipped <- create_pseudobulk_r(m_clipped, cell_group, "sum")
+          m_bpcells_clipped <- pseudobulk_matrix(m_bpcells, cell_group, method = "sum", clip_values = quantile)
+          expect_equal(m_sum_clipped, m_bpcells_clipped)
+        }
+      }
+      expect_equal(m_mean, m_bpcells_mean)
+      expect_equal(m_non_zeros, m_bpcells_non_zeros)
+      # make sure that we dont check for variances if we have number of groups == number of cells
+      if (length(unique(cell_group)) < ncol(m)) {
+        m_var <- create_pseudobulk_r(m, cell_group, "var")
+        m_bpcells_var <- pseudobulk_matrix(m_bpcells, cell_group, method = "variance")
+        expect_equal(m_var, m_bpcells_var, info = paste("Transposed:", m_bpcells@transpose))
+      }
+    }
+  }
+})
+
+test_that("Pseudobulk aggregation works with multiple return types", {
+  m0 <- generate_sparse_matrix(20, 10, max_val = 10)
+  m1 <- m0 |> as("IterableMatrix")
+  m0 <- as.matrix(m0)
+  m0_t <- t(m0)
+  m1_t <- t(m1)
+  methods <- c("nonzeros", "sum", "mean", "variance")
+  for (matrices in list(list(m0_t, m1_t), list(m0, m1))) {
+    # Check across two equal groups, one group, and groups of equal length
+    m <- matrices[[1]]
+    m_bpcells <- matrices[[2]]
+    cell_group <- c(rep.int(1, ncol(m)/2), rep.int(2, ncol(m)/2))
+    m_sum <- create_pseudobulk_r(m, cell_group, "sum")
+    m_mean <- create_pseudobulk_r(m, cell_group, "mean")
+    m_non_zeros <- create_pseudobulk_r(m, cell_group, "non_zeros")
+    m_var <- create_pseudobulk_r(m, cell_group, "var")
+    for (first_method_idx in seq(1, length(methods)-1)) {
+      for (second_method_idx in seq(first_method_idx+1, length(methods))) {
+        methods_of_interest <- c(methods[first_method_idx], methods[second_method_idx])
+        m_bpcells_res <- pseudobulk_matrix(m_bpcells, cell_group, method = methods_of_interest)
+        if ("variance" %in% methods_of_interest) {
+          expect_equal(m_var, m_bpcells_res$var)
+        }
+        if ("mean" %in% methods_of_interest) {
+          expect_equal(m_mean, m_bpcells_res$mean)
+        }
+        if ("sum" %in% methods_of_interest) {
+          expect_equal(m_sum, m_bpcells_res$sum)
+        }
+        if ("nonzeros" %in% methods_of_interest) {
+          expect_equal(m_non_zeros, m_bpcells_res$nonzeros)
+        }
+      }
+    }
+  }
+  # Check that the function errors out when an invalid method is passed
+  expect_error(pseudobulk_matrix(m1, cell_group, method = "nonexistent_method"))
+  expect_error(pseudobulk_matrix(m1, cell_group, method = c("nonexistent_method", "sum")))
```

Comment:
For these `expect_error` tests, it's often helpful to fill in the 2nd argument (regex on what the error message should say). In this case, you did have an error but it was uncontrolled rather than printing out something useful so the tests gave some false confidence.